jax/dtypes.py

```diff
@@ -67,7 +67,8 @@ class _bfloat16_finfo(object):
 # complex_ = np.complex64
 
 # Trivial vectorspace datatype needed for tangent values of int/bool primals
-float0 = np.dtype([('float0', np.void, 0)])
+float0 = np.dtype([('float0', np.bool_, 0)])
+float0 = np.array((0,), float0).dtype
```

Comment:
I don't understand why we need to do this dance.. Anyone know?

Reply:
@shoyer @jakevdp we pray to you for wisdom...

Reply:
This second line creates a differently structured dtype:
```
In [22]: import numpy as np

In [23]: float0 = np.dtype([('float0', np.bool_, 0)])

In [24]: float0_v2 = np.array((0,), float0).dtype

In [25]: float0
Out[25]: dtype([('float0', '?', (0,))])

In [26]: float0_v2
Out[26]: dtype({'names':['float0'], 'formats':[('?', (0,))], 'offsets':[0], 'itemsize':1})
```

I don't know exactly _why_ `dtype` is not idempotent here (it is for the old version of `float0`). My guess is it's a bug in NumPy. As you can imagine, size 0 dtypes are not well tested or used often.

Regardless, I would hesitate to recommend using this version of `float0`, because it does actually allocate 1 bytes of memory per array element:
```
In [3]: np.zeros((10, 10), jax.float0).nbytes
Out[3]: 100
```

Reply:
Maybe it's a bug that's been fixed in recent numpy version?
```python
import numpy as np
print(np.__version__)
float0 = np.dtype([('float0', np.bool_, 0)])
float0_v2 = np.array((0,), float0).dtype
print(float0)
print(float0_v2)
```
```
1.18.5
[('float0', '?', (0,))]
[('float0', '?', (0,))]
```